share/spack/templates/mock-repository/package.pyt

```diff
@@ -0,0 +1,19 @@
+class {{ cls_name }}(Package):
+    homepage = "http://www.example.com"
+    url = "http://www.example.com/root-1.0.tar.gz"
+
+    version("3.0", sha256='abcde')
+    version("2.0", sha256='abcde')
+    version("1.0", sha256='abcde')
+
+{% for dep_spec, dep_type, condition in dependencies %}
+{% if dep_type and condition %}
+    depends_on("{{ dep_spec }}", type="{{ dep_type }}", when="{{ condition }}")
+{% elif dep_type %}
+    depends_on("{{ dep_spec }}", type="{{ dep_type }}")
+{% elif condition %}
+    depends_on("{{ dep_spec }}", when="{{ condition }}")
+{% else %}
+    depends_on("{{ dep_spec }}")
+{% endif %}
+{% endfor %}
```

Comment:
Would the mechanism from https://github.com/spack/spack/pull/27987 also be a suitable replacement? i.e. that PR adds the ability to define packages exactly and in-line with the testing code:

```
_pkgy = (
     "y",
     """\
 class Y(Package):
     version('2.5')
     version('2.4')
     version('2.3', deprecated=True)

     variant('shared', default=True,
             description='Build shared libraries')
 """,
 )
```

I think this direct approach is useful because users know exactly what the package "should be" (e.g. without doing any mental substitution based on the template).

Reply:
I think I like the approach here more, since it is more compact and can be extended further. It also has a single place where these mock packages are collected, so we don't scatter the code around. 

That said, the main point of this change was to get rid of `MockPackageMultiRepo` etc. since that machinery requires maintenance every time we change properties of repositories. Here I remove the mock class and add the possibility to create small repositories to be used during tests on the fly.

Reply:
> I think I like the approach here more, since it is more compact and can be extended further. 

I'm not sure what is more extensible than being able to write the package contents down to the character: a template is flexible but it will never be that flexible. Also, ultimately demands on the template will encourage for example replacing the hardcoded versions that you have there with a placeholder for versions that you insert: at that point, my concern is that you will have created an interface for creating packages which is analogous to (and redundant with) the package.py format.

> It also has a single place where these mock packages are collected, so we don't scatter the code around.

I'm not sure what you mean here: `test_save_dependency_spec_jsons_subset` for example creates its own repo as part of the test, that seems equally as disperse as the preceding approach.

Additionally, IMO consolidation of mock packages is problematic: I see in many cases that `builtin.mock` packages are used for multiple tests and thereby aggregate unrelated behaviors that makes them hard to understand (e.g. in the process of understanding a unit test, a user tracks down the associated `builtin.mock` package and reads it, and then must parse out the components of that package which happen to relate to the specific test).

Reply:
> I'm not sure what is more extensible than being able to write the package contents down to the character: a template is flexible but it will never be that flexible.

This is not true, since writing a literal package can't possibly deal with dynamic content like:
https://github.com/spack/spack/blob/97bd867518c860c16acb85cf601fd4aa2c24e91c/lib/spack/spack/concretize.py#L753-L774

where the package to be generated depends on the function arguments. I can, of course, concatenate strings etc. but at that point I would have re-created a basic template engine. That said my point is:
- Before this PR we were mocking a `Package` without having a `package.py` available and that was brittle[^1]
- In #27987 we introduced a mechanism to use a literal `package.py` in tests. This is fine with me, and I don't want to change that in this PR.
- Here I introduce a class that can create a mock repository with mock `package.py` from API. That is flexible enough to be used in many places where we had similar code before (hence the saving of roughly 200 lines)

> Additionally, IMO consolidation of mock packages is problematic:

To be clear, I don't plan to consolidate `builtin.mock` but only the tests that are using `MockPackage`

[^1]: Changing a few signatures caused tests relying on the mock to fail, and solving that was not straightforward.